content/02_KeyTypes.md

```diff
@@ -1,488 +1,627 @@
+
+So as long as you're using Rx to build the observable sources you need, you get to depend on these rules in your callbacks that receive the events, and it's mostly Rx's problem to keep to the rules. So the main effect of these rules is that it makes life simpler for code that consumes events.
+
+These rules are sometimes expressed as a _grammar_. For example, consider this regular expression:
+
+```
+(OnNext)*(OnError|OnComplete)
+```
+
+This formally captures the basic idea: there can be any number of calls to `OnNext` (maybe even zero calls), that occur in sequence, followed by either an `OnError` or an `OnComplete`, but not both, and there must be nothing after either of these.
+
+One last point: sequences may be infinite. This is true for `IEnumerable<T>`—it's perfectly possible for an enumerator to return `true` every time `MoveNext` is returned, in which case a `foreach` loop iterating over it will never reach the end. It might choose to stop (with a `break` or `return`), or some exception that did not originate from the enumerator might cause the loop to terminate, but it's absolutely acceptable for an `IEnumerable<T>` to produce items for as long as you keep asking for them. The same is true of a `IObservable<T>`—if you subscribe to an observable source, and by the time your program exits you've not received a call to either `OnComplete` or `OnError`, that's not a bug.
+
+So you might argue that this is a slightly better way to describe the rules formally:
+
+```
+(OnNext)*(OnError|OnComplete)?
+```
+
+More subtly, observable sources are allowed to do nothing at all. In fact there's a built-in implementation to save developers from the effort of writing a source that does nothing: if you call `Observable.Never<int>()` it will return an `IObservable<int>`, and if you subscribe to that, it will never call any methods on your observer. This might not look immediately useful—it is logically equivalent to an `IEnumerable<T>` in which the enumerator's `MoveNext` method never returns, which might not be usefully distinguishable from crashing. It's slightly different with Rx, because when we model this "no items emerge ever" behaviour, we don't need to block a thread forever to do it. We can just decide never to call any methods on the observer. This may seem daft, but as you've seen with the `Quiescent` example, sometimes we create observable sources not because we want the actual items that emerge from it, but because we're interested in the instants when interesting things happen. It can sometimes be useful to be able to model "nothing interesting ever happens" cases.
```

Comment:
I have tried. It's surprisingly hard, because in practice these things are almost buried somewhere at the deepest level of something like the `Quiescent` example, at which point they're very hard to relate to anything real. So what I've come up with is a concrete but made up example masquerading as a real world one.